pymeasure/display/widgets.py

```diff
@@ -418,3 +422,403 @@ def update_plot(self, filename):
+            parameter = self.names_inv[name]
+            values = self.eval_string(
+                self.tree.itemWidget(item, 2).text(),
+                name, depth,
+            )
+
+            try:
+                sequence_entry = [{parameter: value} for value in values]
+            except TypeError:
+                log.error(
+                    "TypeError, likely no sequence for one of the parameters"
+                )
+            else:
+                current_sequence[depth].extend(sequence_entry)
+
+            iterator += 1
+            next_depth = self._depth_of_child(iterator.value())
+
+            for depth_idx in range(depth, next_depth, -1):
+                temp_sequence[depth_idx].extend(current_sequence[depth_idx])
+
+                if depth_idx != 0:
+                    sequence_products = list(product(
+                        current_sequence[depth_idx - 1],
+                        temp_sequence[depth_idx]
+                    ))
+
+                    try:
+                        element = sequence_products[0][1]
+                    except IndexError:
+                        log.error(
+                            "IndexError, likely empty nested parameter"
+                        )
+                    else:
+                        if isinstance(element, tuple):
+                            sequence_products = [
+                                (v1, *v2) for (v1, v2) in sequence_products
+                            ]
+
+                    temp_sequence[depth_idx - 1].extend(sequence_products)
+                    temp_sequence[depth_idx] = []
+
+                current_sequence[depth_idx] = []
+                current_sequence[depth_idx - 1] = []
+
+        sequences = temp_sequence[0]
+
+        for idx in range(len(sequences)):
+            if not isinstance(sequences[idx], tuple):
+                sequences[idx] = (sequences[idx],)
+
+        return sequences
+
+    @staticmethod
+    def _depth_of_child(item):
+        """
+        Determine the level / depth of a child item in the sequence tree.
+        """
+
+        depth = -1
+        while item:
+            item = item.parent()
+            depth += 1
+        return depth
+
+    @staticmethod
+    def eval_string(string, name=None, depth=None):
+        """
+        Evaluate the given string. The string is evaluated using a list of
+        pre-defined functions that are deemed safe to use, to prevent the
+        execution of malicious code. For this purpose, also any built-in
+        functions or global variables are not available.
+
+        :param string: String to be interpreted.
+        :param name: Name of the to-be-interpreted string, only used for
+            error messages.
+        :param depth: Depth of the to-be-interpreted string, only used
+            for error messages.
+        """
+
+        evaluated_string = None
+        if len(string) > 0:
+            try:
+                evaluated_string = eval(
```

Comment:
I think this is a good use of `eval`. I think this is a safe usage because you are blocking globals and setting the locals to the numpy functions.